File: py/tntnode.py
import os
import re
import time
import json
from datetime import datetime
from pathlib import Path
import asyncio
import requests
from websocket import WebSocketApp, WebSocketException
from uuid import uuid4
from fake_useragent import UserAgent
from dotenv import load_dotenv
from colorama import init, Fore, Style

init()
# logging
COLORS = {
    'reset': Style.RESET_ALL,
    'cyan': Fore.CYAN,
    'green': Fore.GREEN,
    'yellow': Fore.YELLOW,
    'red': Fore.RED,
    'white': Fore.WHITE,
    'bold': Style.BRIGHT,
    'blue': Fore.BLUE,
    'magenta': Fore.MAGENTA,
    'gray': Fore.LIGHTBLACK_EX,
}

# Fungsi logger
class Logger:
    @staticmethod
    def info(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['cyan']}[INFO]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def warn(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['yellow']}[WARN]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def error(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['red']}[ERROR]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def success(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['green']}[SUCCESS]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def loading(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['magenta']}[LOADING]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def step(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['blue']}[STEP]{prefix} {COLORS['bold']}{msg}{COLORS['reset']}")

    @staticmethod
    def critical(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['red']}{COLORS['bold']}[FATAL]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def summary(msg):
        print(f"{COLORS['green']}{COLORS['bold']}[SUMMARY] {msg}{COLORS['reset']}")

    @staticmethod
    def bandwidth(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"{COLORS['white']}[BANDWIDTH]{prefix} {msg}{COLORS['reset']}")

    @staticmethod
    def banner():
        banner = f"""
{COLORS['green']}============================ WELCOME TO TITAN NODE ============================={COLORS['reset']}
{COLORS['yellow']}
 ██████╗██╗   ██║ █████╗ ███╗   ██║███╗   ██║ ██████╗ ██████╗ ███████╗
██╔════╝██║   ██║██╔══██╗████╗  ██║████╗  ██║██╔═══██╗██╔══██╗██╔════╝
██║     ██║   ██║███████║██╔██╗ ██║██╔██╗ ██║██║   ██║██║  ██║█████╗  
██║     ██║   ██║██╔══██║██║╚██╗██║██║╚██╗██║██║   ██║██║  ██║██╔══╝  
╚██████╗╚██████╔╝██║  ██║██║ ╚████║██║ ╚████║╚██████╔╝██████╔╝███████╗
 ╚═════╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝╚═╝  ╚═══╝ ╚═════╝ ╚═════╝ ╚══════╝
{COLORS['reset']}
{COLORS['cyan']}================== TitanNet Node Runner by CUANNODE =================={COLORS['reset']}
{COLORS['magenta']}       Powered by Insider - Run Multiple Nodes with Ease       {COLORS['reset']}
{COLORS['yellow']}           Credit: Greyscope&Co, Arcxteam, xAI Community       {COLORS['reset']}
{COLORS['cyan']}======================================================================={COLORS['reset']}
"""
        print(banner)

    @staticmethod
    def section(msg, account_index=None):
        line = '─' * 50
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"\n{COLORS['gray']}{line}{COLORS['reset']}")
        if msg:
            print(f"{COLORS['white']}{COLORS['bold']}{prefix} {msg} {COLORS['reset']}")
        print(f"{COLORS['gray']}{line}{COLORS['reset']}\n")

    @staticmethod
    def countdown(msg, account_index=None):
        prefix = f" Account {account_index}:" if account_index else ""
        print(f"\r{COLORS['blue']}[COUNTDOWN]{prefix} {msg}{COLORS['reset']}", end="")

# proxies.txt
def read_proxies():
    proxy_file_path = Path(__file__).parent / 'proxies.txt'
    try:
        if proxy_file_path.exists():
            with open(proxy_file_path, 'r', encoding='utf-8') as f:
                proxies = [p.strip() for p in f.readlines() if p.strip() and re.match(r'^http(s)?:\/\/[^:]+(:\w+)?@[^:]+:\d+$', p.strip())]
            if not proxies:
                Logger.warn('No valid proxies found in proxies.txt.')
            return proxies
        Logger.warn('proxies.txt not found.')
        return []
    except Exception as e:
        Logger.error(f"Error reading proxies.txt: {str(e)}")
        return []

# .env
def read_refresh_tokens():
    load_dotenv()
    tokens = []
    i = 1
    while os.getenv(f'REFRESH_TOKEN_{i}'):
        tokens.append(os.getenv(f'REFRESH_TOKEN_{i}'))
        i += 1
    if not tokens and os.getenv('REFRESH_TOKEN'):
        tokens.append(os.getenv('REFRESH_TOKEN'))
    if not tokens:
        Logger.error('No refresh tokens found in .env file.')
    return tokens

class TitanNode:
    def __init__(self, refresh_token, proxy=None, account_index=1, proxy_index=None):
        self.refresh_token = refresh_token
        self.proxy = proxy
        self.account_index = account_index
        self.proxy_index = proxy_index
        self.access_token = None
        self.user_id = None
        self.email = None
        self.device_id = str(uuid4())
        self.data_sent = 0
        self.data_received = 0
        self.servers = [
            {'api': 'https://task.titannet.info', 'ws': 'wss://task.titannet.info'},
            {'api': 'https://task.titanedge.cn', 'ws': 'wss://task.titanedge.cn'},
            {'api': 'https://task.titannet.io', 'ws': 'wss://task.titannet.io'},
            {'api': 'https://task.titandev.info', 'ws': 'wss://task.titandev.info'},
        ]
        self.current_server_index = 0
        self.session = requests.Session()
        self.ua = UserAgent()
        self.session.headers.update({
            'Accept': '*/*',
            'Accept-Language': 'en-US,en;q=0.9',
            'Content-Type': 'application/json',
            'User-Agent': self.ua.random,
        })
        if self.proxy:
            self.session.proxies = {'http': self.proxy, 'https': self.proxy}
        self.ws = None
        self.reconnect_interval = 5 * 60  # 5 menit
        self.ping_interval = None
        self.reconnect_attempts = 0
        self.max_reconnect_attempts = 5
        self.is_single_account = None  # Akan diatur di start()

    async def refresh_access_token(self):
        Logger.loading("Refreshing access token...", None if self.is_single_account else self.account_index)
        try:
            response = self.session.post(
                f"{self.servers[self.current_server_index]['api']}/api/auth/refresh-token",
                json={"refresh_token": self.refresh_token}
            )
            self.data_sent += len(json.dumps({"refresh_token": self.refresh_token}))
            self.data_received += len(json.dumps(response.json()))
            data = response.json()
            if data.get('code') == 0:
                self.access_token = data['data']['access_token']
                self.user_id = data['data']['user_id']
                self.email = data['data'].get('email', 'Unknown')
                self.session.headers.update({'Authorization': f'Bearer {self.access_token}'})
                Logger.success(f"Access token refreshed. Email: {self.email}", None if self.is_single_account else self.account_index)
                Logger.bandwidth(f"Data Sent: {self.data_sent} bytes, Received: {self.data_received} bytes", None if self.is_single_account else self.account_index)
                return True
            else:
                Logger.error(f"Failed to refresh token: {data.get('msg', 'Unknown error')}", None if self.is_single_account else self.account_index)
                return False
        except Exception as e:
            Logger.error(f"Error refreshing access token: {str(e)}", None if self.is_single_account else self.account_index)
            return False

    async def register_node(self):
        Logger.loading("Registering node...", None if self.is_single_account else self.account_index)
        try:
            payload = {
                "ext_version": "0.0.4",
                "language": "en",
                "user_script_enabled": True,
                "device_id": self.device_id,
                "install_time": datetime.utcnow().isoformat(),
            }
            response = self.session.post(
                f"{self.servers[self.current_server_index]['api']}/api/webnodes/register",
                json=payload
            )
            self.data_sent += len(json.dumps(payload))
            self.data_received += len(json.dumps(response.json()))
            data = response.json()
            if data.get('code') == 0:
                Logger.success("Node registered successfully.", None if self.is_single_account else self.account_index)
                Logger.success(f"Initial Points: {json.dumps(data['data'])}", None if self.is_single_account else self.account_index)
            else:
                Logger.error(f"Node registration failed: {data.get('msg', 'Unknown error')}", None if self.is_single_account else self.account_index)
        except Exception as e:
            Logger.error(f"Error registering node: {str(e)}", None if self.is_single_account else self.account_index)

    def connect_websocket(self):
        server = self.servers[self.current_server_index]
        Logger.loading(f"Connecting to WebSocket at {server['ws']}...", None if self.is_single_account else self.account_index)
        ws_url = f"{server['ws']}/api/public/webnodes/ws?token={self.access_token}&device_id={self.device_id}"
        headers = {'User-Agent': self.session.headers['User-Agent']}
        proxy_dict = {'http': self.proxy, 'https': self.proxy} if self.proxy else None

        def on_open(ws):
            Logger.success(f"WebSocket connection established to {server['ws']}. Waiting for jobs...", None if self.is_single_account else self.account_index)
            self.reconnect_attempts = 0
            async def ping():
                while self.ws and self.ws.sock and self.ws.sock.connected:
                    echo_message = json.dumps({"cmd": 1, "echo": "echo me", "jobReport": {"cfgcnt": 2, "jobcnt": 0}})
                    self.ws.send(echo_message)
                    self.data_sent += len(echo_message)
                    await asyncio.sleep(30)
            asyncio.ensure_future(ping())

        def on_message(ws, data):
            try:
                self.data_received += len(data)
                message = json.loads(data)
                if message.get('cmd') == 1:
                    response = json.dumps({"cmd": 2, "echo": message['echo']})
                    ws.send(response)
                    self.data_sent += len(response)
                if message.get('userDataUpdate'):
                    Logger.success(
                        f"TNTIP Points Update - Get Today: {message['userDataUpdate']['today_points']}, Get Total: {message['userDataUpdate']['total_points']}",
                        None if self.is_single_account else self.account_index
                    )
                Logger.bandwidth(f"Data Sent: {self.data_sent} bytes, Received: {self.data_received} bytes", None if self.is_single_account else self.account_index)
            except Exception as e:
                Logger.warn(f"Could not parse message: {data}", None if self.is_single_account else self.account_index)

        def on_error(ws, error):
            Logger.error(f"WebSocket error: {str(error)}", None if self.is_single_account else self.account_index)
            ws.close()

        def on_close(ws, close_status_code, close_msg):
            Logger.warn("WebSocket connection closed. Attempting to reconnect...", None if self.is_single_account else self.account_index)
            if self.ping_interval:
                self.ping_interval.cancel()
            if self.reconnect_attempts < self.max_reconnect_attempts:
                self.reconnect_attempts += 1
                delay = self.reconnect_interval * (2 ** self.reconnect_attempts)
                Logger.info(f"Reconnecting in {delay} seconds...", None if self.is_single_account else self.account_index)
                time.sleep(delay)
                asyncio.run(self.try_next_server())
            else:
                Logger.critical(f"Max reconnect attempts reached on {server['ws']}. Stopping.", None if self.is_single_account else self.account_index)

        self.ws = WebSocketApp(ws_url, header=headers, on_open=on_open, on_message=on_message, on_error=on_error, on_close=on_close)
        self.ws.run_forever(proxy=proxy_dict)

    async def try_next_server(self):
        self.current_server_index = (self.current_server_index + 1) % len(self.servers)
        new_server = self.servers[self.current_server_index]
        Logger.info(f"Switching to server: {new_server['api']}", None if self.is_single_account else self.account_index)
        self.session = requests.Session()
        self.session.headers.update({
            'Accept': '*/*',
            'Accept-Language': 'en-US,en;q=0.9',
            'Content-Type': 'application/json',
            'User-Agent': self.ua.random,
        })
        if self.proxy:
            self.session.proxies = {'http': self.proxy, 'https': self.proxy}
        await self.start()

    async def start(self):
        self.is_single_account = len(read_refresh_tokens()) == 1
        Logger.banner()
        Logger.section(f"Starting Account {self.account_index}", None if self.is_single_account else self.account_index)
        Logger.step(f"Device ID: {self.device_id}", None if self.is_single_account else self.account_index)
        if self.proxy:
            Logger.info(f"Using Proxy {self.proxy_index + 1}: {self.proxy}", None if self.is_single_account else self.account_index)
        else:
            Logger.info("Running in Direct Mode (No Proxy)", None if self.is_single_account else self.account_index)

        if await self.refresh_access_token():
            await self.register_node()
            loop = asyncio.get_event_loop()
            loop.run_in_executor(None, self.connect_websocket)
        else:
            Logger.error("Could not start bot due to token refresh failure.", None if self.is_single_account else self.account_index)
            if self.current_server_index < len(self.servers) - 1:
                await self.try_next_server()

async def main():
    refresh_tokens = read_refresh_tokens()
    proxies = read_proxies()

    if not refresh_tokens:
        Logger.critical("No valid refresh tokens found. Please add REFRESH_TOKEN or REFRESH_TOKEN_1, etc., to .env file.")
        return

    if len(refresh_tokens) > 1 and len(proxies) < len(refresh_tokens):
        Logger.critical("Multiple accounts require an equal or greater number of proxies. Please add more proxies to proxies.txt.")
        return

    Logger.summary(f"Starting {len(refresh_tokens)} account(s) with {len(proxies)} available proxies.")

    tasks = []
    for index, token in enumerate(refresh_tokens):
        proxy = proxies[index % len(proxies)] if len(refresh_tokens) > 1 else None
        proxy_index = index % len(proxies) if len(refresh_tokens) > 1 else None
        node = TitanNode(token, proxy, index + 1, proxy_index)
        tasks.append(node.start())
        await asyncio.sleep(15)  # Delay 15

    await asyncio.gather(*tasks)

if __name__ == "__main__":
    asyncio.run(main())

File: js/tntnode.js
require('dotenv').config();
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const WebSocket = require('ws');
const { v4: uuidv4 } = require('uuid');
const { HttpsProxyAgent } = require('https-proxy-agent');
const randomUseragent = require('random-useragent');

const colors = {
    reset: "\x1b[0m",
    cyan: "\x1b[36m",
    green: "\x1b[32m",
    yellow: "\x1b[33m",
    red: "\x1b[31m",
    white: "\x1b[37m",
    bold: "\x1b[1m",
    blue: "\x1b[34m",
    magenta: "\x1b[35m",
    gray: "\x1b[90m",
};

const logger = {
    info: (msg, accountIndex = null) => console.log(`${colors.cyan}[INFO]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    warn: (msg, accountIndex = null) => console.log(`${colors.yellow}[WARN]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    error: (msg, accountIndex = null) => console.log(`${colors.red}[ERROR]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    success: (msg, accountIndex = null) => console.log(`${colors.green}[SUCCESS]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    loading: (msg, accountIndex = null) => console.log(`${colors.magenta}[LOADING]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    step: (msg, accountIndex = null) => console.log(`${colors.blue}[STEP]${accountIndex ? ` Account ${accountIndex}:` : ''} ${colors.bold}${msg}${colors.reset}`),
    critical: (msg, accountIndex = null) => console.log(`${colors.red}${colors.bold}[FATAL]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    summary: (msg) => console.log(`${colors.green}${colors.bold}[SUMMARY] ${msg}${colors.reset}`),
    bandwidth: (msg, accountIndex = null) => console.log(`${colors.white}[BANDWIDTH]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
    banner: () => {
        const banner = `
${colors.green}============================ WELCOME TO TITAN NODE =============================${colors.reset}
${colors.yellow}
 ██████╗██╗   ██║ █████╗ ███╗   ██║███╗   ██║ ██████╗ ██████╗ ███████╗
██╔════╝██║   ██║██╔══██╗████╗  ██║████╗  ██║██╔═══██╗██╔══██╗██╔════╝
██║     ██║   ██║███████║██╔██╗ ██║██╔██╗ ██║██║   ██║██║  ██║█████╗  
██║     ██║   ██║██╔══██║██║╚██╗██║██║╚██╗██║██║   ██║██║  ██║██╔══╝  
╚██████╗╚██████╔╝██║  ██║██║ ╚████║██║ ╚████║╚██████╔╝██████╔╝███████╗
 ╚═════╝ ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝╚═╝  ╚═══╝ ╚═════╝ ╚═════╝ ╚══════╝
${colors.reset}
${colors.cyan}================== TitanNet Node Runner by CUANNODE ==================${colors.reset}
${colors.magenta}       Powered by Insider - Run Multiple Nodes with Ease       ${colors.reset}
${colors.yellow}           Credit: Greyscope&Co, Arcxteam, xAI Community       ${colors.reset}
${colors.cyan}=======================================================================${colors.reset}
`;
        console.log(banner);
    },
    section: (msg, accountIndex = null) => {
        const line = '─'.repeat(50);
        console.log(`\n${colors.gray}${line}${colors.reset}`);
        if (msg) console.log(`${colors.white}${colors.bold}${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg} ${colors.reset}`);
        console.log(`${colors.gray}${line}${colors.reset}\n`);
    },
    countdown: (msg, accountIndex = null) => process.stdout.write(`\r${colors.blue}[COUNTDOWN]${accountIndex ? ` Account ${accountIndex}:` : ''} ${msg}${colors.reset}`),
};

/**
 * Reads proxies from proxies.txt
 * @returns {string[]}
 */
function readProxies() {
    const proxyFilePath = path.join(__dirname, 'proxies.txt');
    try {
        if (fs.existsSync(proxyFilePath)) {
            const proxies = fs.readFileSync(proxyFilePath, 'utf-8')
                .split('\n')
                .map(p => p.trim())
                .filter(p => p && /^http(s)?:\/\/[^:]+(:\w+)?@[^:]+:\d+$/.test(p));
            if (proxies.length === 0) logger.warn('No valid proxies found in proxies.txt.');
            return proxies;
        }
        logger.warn('proxies.txt not found.');
        return [];
    } catch (error) {
        logger.error(`Error reading proxies.txt: ${error.message}`);
        return [];
    }
}

/**
 * Reads refresh tokens from .env
 * @returns {string[]}
 */
function readRefreshTokens() {
    const tokens = [];
    let i = 1;
    while (process.env[`REFRESH_TOKEN_${i}`]) {
        tokens.push(process.env[`REFRESH_TOKEN_${i}`]);
        i++;
    }
    if (!tokens.length && process.env.REFRESH_TOKEN) {
        tokens.push(process.env.REFRESH_TOKEN);
    }
    if (!tokens.length) {
        logger.error('No refresh tokens found in .env file.');
    }
    return tokens;
}

class TitanNode {
    constructor(refreshToken, proxy = null, accountIndex = 1, proxyIndex = null) {
        this.refreshToken = refreshToken;
        this.proxy = proxy;
        this.accountIndex = accountIndex;
        this.proxyIndex = proxyIndex;
        this.accessToken = null;
        this.userId = null;
        this.email = null;
        this.deviceId = uuidv4();
        this.dataSent = 0;
        this.dataReceived = 0;
        this.servers = [
            { api: 'https://task.titannet.info', ws: 'wss://task.titannet.info' },
            { api: 'https://task.titanedge.cn', ws: 'wss://task.titanedge.cn' },
            { api: 'https://task.titannet.io', ws: 'wss://task.titannet.io' },
            { api: 'https://task.titandev.info', ws: 'wss://task.titandev.info' },
        ];
        this.currentServerIndex = 0;

        const agent = this.proxy ? new HttpsProxyAgent(this.proxy) : null;

        this.api = axios.create({
            baseURL: this.servers[this.currentServerIndex].api,
            httpsAgent: agent,
            headers: {
                'Accept': '*/*',
                'Accept-Language': 'en-US,en;q=0.9',
                'Content-Type': 'application/json',
                'User-Agent': randomUseragent.getRandom(),
            },
        });

        this.api.interceptors.request.use((config) => {
            this.dataSent += JSON.stringify(config.data || {}).length;
            return config;
        });

        this.api.interceptors.response.use((response) => {
            this.dataReceived += JSON.stringify(response.data || {}).length;
            return response;
        });

        this.ws = null;
        this.reconnectInterval = 1000 * 60 * 5;
        this.pingInterval = null;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 5;
        this.isSingleAccount = null; // Will be set in start()
    }

    async refreshAccessToken() {
        logger.loading(`Refreshing access token...`, this.accountIndex);
        try {
            const response = await this.api.post('/api/auth/refresh-token', {
                refresh_token: this.refreshToken,
            });

            if (response.data && response.data.code === 0) {
                this.accessToken = response.data.data.access_token;
                this.userId = response.data.data.user_id;
                this.email = response.data.data.email || 'Unknown';
                this.api.defaults.headers.common['Authorization'] = `Bearer ${this.accessToken}`;
                logger.success(`Access token refreshed. Email: ${this.email}`, this.accountIndex);
                logger.bandwidth(`Data Sent: ${this.dataSent} bytes, Received: ${this.dataReceived} bytes`, this.accountIndex);
                return true;
            } else {
                logger.error(`Failed to refresh token: ${response.data.msg || 'Unknown error'}`, this.accountIndex);
                return false;
            }
        } catch (error) {
            logger.error(`Error refreshing access token: ${error.message}`, this.accountIndex);
            return false;
        }
    }

    async registerNode() {
        logger.loading(`Registering node...`, this.accountIndex);
        try {
            const payload = {
                ext_version: "0.0.4",
                language: "en",
                user_script_enabled: true,
                device_id: this.deviceId,
                install_time: new Date().toISOString(),
            };
            const response = await this.api.post('/api/webnodes/register', payload);

            if (response.data && response.data.code === 0) {
                logger.success(`Node registered successfully.`, this.accountIndex);
                logger.success(`Initial Points: ${JSON.stringify(response.data.data)}`, this.accountIndex);
            } else {
                logger.error(`Node registration failed: ${response.data.msg || 'Unknown error'}`, this.accountIndex);
            }
        } catch (error) {
            logger.error(`Error registering node: ${error.message}`, this.accountIndex);
        }
    }

    connectWebSocket() {
        const server = this.servers[this.currentServerIndex];
        logger.loading(`Connecting to WebSocket at ${server.ws}...`, this.accountIndex);
        const wsUrl = `${server.ws}/api/public/webnodes/ws?token=${this.accessToken}&device_id=${this.deviceId}`;
        const agent = this.proxy ? new HttpsProxyAgent(this.proxy) : null;

        this.ws = new WebSocket(wsUrl, {
            agent: agent,
            headers: {
                'User-Agent': this.api.defaults.headers['User-Agent'],
            },
        });

        this.ws.on('open', () => {
            logger.success(`WebSocket connection established to ${server.ws}. Waiting for jobs...`, this.accountIndex);
            this.reconnectAttempts = 0;
            this.pingInterval = setInterval(() => {
                if (this.ws.readyState === WebSocket.OPEN) {
                    const echoMessage = JSON.stringify({ cmd: 1, echo: "echo me", jobReport: { cfgcnt: 2, jobcnt: 0 } });
                    this.ws.send(echoMessage);
                    this.dataSent += echoMessage.length;
                }
            }, 30 * 1000);
        });

        this.ws.on('message', (data) => {
            try {
                this.dataReceived += data.length;
                const message = JSON.parse(data);
                if (message.cmd === 1) {
                    const response = JSON.stringify({ cmd: 2, echo: message.echo });
                    this.ws.send(response);
                    this.dataSent += response.length;
                }
                if (message.userDataUpdate) {
                    logger.success(`TNTIP Points Update - Get Today: ${message.userDataUpdate.today_points}, Get Total: ${message.userDataUpdate.total_points}`, this.accountIndex);
                }
                logger.bandwidth(`Data Sent: ${this.dataSent} bytes, Received: ${this.dataReceived} bytes`, this.accountIndex);
            } catch (error) {
                logger.warn(`Could not parse message: ${data}`, this.accountIndex);
            }
        });

        this.ws.on('error', (error) => {
            logger.error(`WebSocket error: ${error.message}`, this.accountIndex);
            this.ws.close();
        });

        this.ws.on('close', () => {
            logger.warn(`WebSocket connection closed. Attempting to reconnect...`, this.accountIndex);
            clearInterval(this.pingInterval);
            if (this.reconnectAttempts < this.maxReconnectAttempts) {
                this.reconnectAttempts++;
                const delay = this.reconnectInterval * Math.pow(2, this.reconnectAttempts);
                setTimeout(() => this.tryNextServer(), delay);
            } else {
                logger.critical(`Max reconnect attempts reached on ${server.ws}. Stopping.`, this.accountIndex);
            }
        });
    }

    tryNextServer() {
        this.currentServerIndex = (this.currentServerIndex + 1) % this.servers.length;
        const newServer = this.servers[this.currentServerIndex];
        logger.info(`Switching to server: ${newServer.api}`, this.accountIndex);
        this.api.defaults.baseURL = newServer.api;
        this.start();
    }

    async start() {
        this.isSingleAccount = readRefreshTokens().length === 1;
        logger.banner();
        logger.section(`Starting Account ${this.accountIndex}`, this.isSingleAccount ? null : this.accountIndex);
        logger.step(`Device ID: ${this.deviceId}`, this.isSingleAccount ? null : this.accountIndex);
        if (this.proxy) {
            logger.info(`Using Proxy ${this.proxyIndex + 1}: ${this.proxy}`, this.isSingleAccount ? null : this.accountIndex);
        } else {
            logger.info(`Running in Direct Mode (No Proxy)`, this.isSingleAccount ? null : this.accountIndex);
        }

        const tokenRefreshed = await this.refreshAccessToken();
        if (tokenRefreshed) {
            await this.registerNode();
            this.connectWebSocket();
        } else {
            logger.error(`Could not start bot due to token refresh failure.`, this.isSingleAccount ? null : this.accountIndex);
            if (this.currentServerIndex < this.servers.length - 1) {
                this.tryNextServer();
            }
        }
    }
}

function main() {
    const refreshTokens = readRefreshTokens();
    const proxies = readProxies();

    if (!refreshTokens.length) {
        logger.critical('No valid refresh tokens found. Please add REFRESH_TOKEN or REFRESH_TOKEN_1, etc., to .env file.');
        return;
    }

    if (refreshTokens.length > 1 && proxies.length < refreshTokens.length) {
        logger.critical('Multiple accounts require an equal or greater number of proxies. Please add more proxies to proxies.txt.');
        return;
    }

    logger.summary(`Starting ${refreshTokens.length} account(s) with ${proxies.length} available proxies.`);

    refreshTokens.forEach((token, index) => {
        const proxy = refreshTokens.length > 1 ? proxies[index % proxies.length] : null;
        const proxyIndex = refreshTokens.length > 1 ? index % proxies.length : null;
        setTimeout(() => {
            const bot = new TitanNode(token, proxy, index + 1, proxyIndex);
            bot.start();
        }, index * 10000);
    });
}

main();